### PR TITLE
storegateway: Detect OOM During Eagerly Loading Index Header

### LINKED ISSUE #4762: store-gateway: retain lazy-loaded index headers between restarts
### Context

The store-gateway can lazily load the [index header](https://grafana.com/docs/mimir/latest/operators-guide/architecture/binary-index-header/) of a block when each block is requested by the querier. The store-gateway loses these loaded index headers upon restart and also unloads them when they haven't been queried after time idle period (3h by default). Loading one index header can take between seconds and minutes.

Related to https://github.com/grafana/mimir/issues/4763

### Problem

When the store-gateway crashes, is rescheduled on a new node or rolled out with a new version it loses the index headers. This means that subsequent queries for the blocks of these index headers will suffer a latency increase.

### Proposal

Retain the list of lazily-loaded index headers on disk and load the non-lazily after a restart. 

* We can periodically persist the list of headers (@ 1 min) to a file on disk. Every minute we overwrite the file with the current list of block IDs. The list of lazily-loaded index headers is kept [here](https://github.com/grafana/mimir/blob/1820c4164f4d5a3c8815c3b0fa848827210e7c6b/pkg/storegateway/indexheader/reader_pool.go#L51).
	* When persisting the file to disk it might be a good idea to also checksum it so that we can recover from forced shutdowns.
	* It might be useful to also persist the time of the last access to the index header so that we can unload idle index headers. Otherwise, frequent restarts may mean effectively disabling lazy-loading.
* Upon startup the store-gateway reads the file and load the index headers that are listed in the file
	* The existing concurrency controls (`blocks-storage.bucket-store.meta-sync-concurrency`, its tenant and its block equivalents) should probably apply in this loading as well
* The store-gateway doesn't pass its readiness check until it has loaded all the previously loaded index headers.

### Alternatives

* https://github.com/grafana/mimir/issues/4763
* Disable lazy-loading. It is less desirable because
	* it increases memory usage since the index header has some overhead while it is loaded (holding every 32nd postings offset table entry)
	* and also increases startup time
```[tasklist]
### Tasks
- [x] Store-Gateway: Periodically persist list of headers to disk and set last update time #5095
- [x] Store-Gateway: Read the list of headers file during startup and update readiness check to wait for this #5596
- [x] Store-Gateway: Take care possible OOM when loading index headers #5644
```

### PR DESCRIPTION
**Pull request by lamida**
**Note:** This PR is purposely targeting `lamida/sg-eager-load-headers` (#5596) for easier diffs, later the target will be changed to `lamida/sg-lazy-load` branch once #5596 is merged. Eventually, all lazy loading PRs will be merged to `main` branch in #5606.

#### What this PR does

[Detect OOM](https://github.com/grafana/mimir/issues/4762#issuecomment-1512854593) in store gateway during eagerly loading lazy loaded index headers in startup. We are writing a marker file when store gateway starting to do eager loading and removing the marker once eager loading runs successfully. If there is a marker file at the beginning of eager loading, that could indicate previous eager loading failed, probably by OOM, and we skip eager loading the index header for such cases.

As part of this PR, I also moved `Exists` method from `shutdownmarker` package to `atomicfs` package because the later package seems more appropriate for the general use of the method.

#### Which issue(s) this PR fixes or relates to

Fixes #4762

#### Checklist

- [X] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storegateway/indexheader/reader_pool.go [hunk 1]
```diff
@@ -214,6 +216,34 @@ func (p *ReaderPool) NewBinaryReader(ctx context.Context, logger log.Logger, bkt
 	return reader, err
 }
 
+func (p *ReaderPool) tryEagerLoadIndexHeadersSnapshot(lazyBinaryReader *LazyBinaryReader, dir string, blockID ulid.ULID) {
```
Comment (charleskorn) [inline pkg/storegateway/indexheader/reader_pool.go:219]: I think we need to handle failing to eagerly load index headers at a higher level. 

If I understand correctly, this will create and delete the marker file once per block, and that marker is scoped per-tenant. 

However, if store-gateway crashes while eagerly loading index headers (eg. because we OOMed), I'm thinking it's best to not eagerly load any index headers from any tenant when the store-gateway restarts. 

With the current setup, we'll only skip the tenants that were being loaded at the time of the crash, and still try to eagerly load all other tenants. However, imagine a scenario where the first tenant eagerly loads X GB, but the next tenant loads just enough to push the store-gateway over its memory limit, causing it to OOM and be killed. On the next run, we'll still load the first tenant's index headers, and only skip the second tenant.

**Comment by lamida** [comment]
@charleskorn before I continue to address your comment in this PR, last week, I had discussion with @dimitarvdimitrov on the need to have this store gateway crash handler.

TLDR: We might not need this PR. 

It is because in #5596 we will remove the snapshot file after loading the block list from the json file. Only after that we will eager load all blocks listed in the json file. If one of block eager loading caused store gateway to crash, after restart, the json file is already gone, hence no more eager loading.

**Comment by charleskorn** [comment]
> @charleskorn before I continue to address your comment in this PR, last week, I had discussion with @dimitarvdimitrov on the need to have this store gateway crash handler.
> 
> TLDR: We might not need this PR.
> 
> It is because in #5596 we will remove the snapshot file after loading the block list from the json file. Only after that we will eager load all blocks listed in the json file. If one of block eager loading caused store gateway to crash, after restart, the json file is already gone, hence no more eager loading.

That sounds great to me, much simpler 🙂

**Comment by lamida** [comment]
Closing this as the discussion above.
